src/spiProcessor/lombok/spi/SpiProcessor.java

```diff
@@ -0,0 +1,254 @@
+	private SpiProcessorCollector data;
+	private SpiProcessorPersistence persistence;
+	
+	static String getRootPathOfServiceFiles() {
+		return "META-INF/services/";
+	}
+	
+	@Override public SourceVersion getSupportedSourceVersion() {
+		return SourceVersion.latest();
+	}
+	
+	static String toErrorMsg(Exception e, String pathName) {
+		StringBuilder sb = new StringBuilder();
+		sb.append("Exception applying SPI processor on ").append(pathName).append(": ").append(toStringLong(e));
+		return sb.toString();
+	}
+	
+	private static String toStringLong(Throwable t) {
+		if (t == null) return "NULL";
+		StringBuilder out = new StringBuilder();
+		
+		String msg = t.getMessage();
+		out.append(t.getClass().getName());
+		if (msg != null) out.append(": ").append(msg);
+		String indent = "  ";
+		
+		StackTraceElement[] elems = t.getStackTrace();
+		if (elems != null) for (int i = 0; i < elems.length; i++) {
+			out.append("\n").append(indent).append(elems[i]);
+		}
+		Throwable cause = t.getCause();
+		while (cause != null) {
+			indent = indent + " ";
+			out.append("\n").append(indent).append("Caused by: ").append(cause.getClass().getName());
+			msg = cause.getMessage();
+			if (msg != null) out.append(": ").append(msg);
+			elems = cause.getStackTrace();
+			indent = indent + " ";
+			if (elems != null) for (int i = 0; i < elems.length; i++) {
+				out.append("\n").append(indent).append(elems[i]);
+			}
+			cause = cause.getCause();
+		}
+		return out.toString();
+	}
+	
+	@Override public synchronized void init(ProcessingEnvironment processingEnv) {
+		super.init(processingEnv);
+		persistence = new SpiProcessorPersistence("SpiProcessor", processingEnv.getFiler(), processingEnv.getMessager());
+		data = new SpiProcessorCollector(processingEnv);
+		for (String serviceName : persistence.tryFind()) data.getService(serviceName);
+		data.stripProvidersWithoutSourceFile();
+	}
+	
+	@Override public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
+		removeStaleData(roundEnv);
+		handleAnnotations(roundEnv);
+		if (roundEnv.processingOver()) writeData();
+		
+		return false;
+	}
+	
+	private void writeData() {
+		for (SpiProcessorService service : data.services()) {
+			try {
+				persistence.write(service.getName(), service.toProvidersListFormat());
+			} 
+			catch (IOException e) {
+				processingEnv.getMessager().printMessage(Kind.ERROR, e.getMessage());
+			}
+		}
+	}
+	
+	private void removeStaleData(RoundEnvironment roundEnv) {
+		for (Element e : roundEnv.getRootElements()) {
+			if (e instanceof TypeElement) {
+				TypeElement currentClass = (TypeElement) e;
+				data.removeProvider(createProperQualifiedName(currentClass));
+			}
+		}
+	}
+	
+	private void handleAnnotations(RoundEnvironment roundEnv) {
+		Set<? extends Element> elements = roundEnv.getElementsAnnotatedWith(Provides.class);
+		for (Element e : elements) handleProvidingElement(e);
+	}
+	
+	private void handleProvidingElement(Element element) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 29. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.